### PR TITLE
Mob following

### PR DESCRIPTION
**Pull request by stevesoltys**
Add support for mob following. Currently a WIP.
- Support for Mobs that have a size greater than one tile is still untested, but _should_ work. See `get_minimum_distance` in `FollowingModule`.
- Pathfinding seemingly needs some work, as it's easy to get a `Mob` stuck behind a wall, even while using the `AStarPathfindingAlgorithm` implementation.
- Support for projectiles is also not there, but will be a simple addition once projectile clipping has been completed. See `follow` in `FollowingModule`.

### CODE REVIEW

File: game/src/main/org/apollo/game/model/Position.java [hunk 1]
```diff
@@ -224,6 +224,66 @@ public boolean isWithinDistance(Position other, int distance) {
...
 	}
 
+	/**
+	 * Gets a position, based upon this position, from the given deltas.
+	 *
+	 * @param deltaX The delta x.
+	 * @param deltaY The delta y.
+	 * @return The position.
+	 */
+	public Position fromDeltas(int deltaX, int deltaY) {
```
Comment (garyttierney) [inline game/src/main/org/apollo/game/model/Position.java]:
```
public Position add(int x, int y) {
```

File: game/src/main/org/apollo/game/model/Position.java [hunk 2]
```diff
@@ -224,6 +224,66 @@ public boolean isWithinDistance(Position other, int distance) {
...
+		return new Position(getX() + deltaX, getY() + deltaY, getHeight() + deltaZ);
+	}
+
+	/**
+	 * Gets the adjacent position in the given direction.
+	 *
+	 * @param direction the direction
+	 * @return the position
+	 */
+	public Position fromDirection(Direction direction) {
```
Comment (garyttierney) [inline game/src/main/org/apollo/game/model/Position.java]:
```
public Position add(Direction direction) {
```
Also use the helper functions in Direction: https://github.com/apollo-rsps/apollo/blob/bugfix/collision-detection/game/src/main/org/apollo/game/model/Direction.java#L190

File: data/plugins/follow/util.rb [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+  def self.follow(mob, other_mob, maximum_distance, simple, projectile_clipped)
+
+    mob.following = other_mob.index
+    mob.interacting_mob = other_mob
+
+    schedule 0, true do |task|
+
+      # if we're following a different entity, stop this task
+      # TODO: Could the indices of NPCs and Players collide since they are in different instances of MobRepository..?
+      if mob.following != other_mob.index
```
Comment (garyttierney) [inline data/plugins/follow/util.rb]: This is a valid point. You might want to store a reference on the Mob rather than an index as an attribtue.

**Comment by WizardJesse1** [comment]
http://recordit.co/Lf1H9RJktV any idea why its always trying to go to the side of the player ;o?

**Comment by garyttierney** [comment]
@WizardJesse1 I believe this is in reference to that:

```
21:09 < shiver> it's not identical to RS atm tho
21:09 < shiver> like, it doesn't do the whole dance thing
21:09 < sfix> o
21:09 < sfix> is it still a WIP?
21:09 < shiver> i have the function for larger entities too
21:09 < shiver> but i haven't tested it
21:09 < sfix> that's a prereq for merging
21:09 < shiver> yea, but it works. maybe merge it but keep the issue?
21:10 < sfix> would rather wait until the feature is complete
21:10 < sfix> but anyway make the PR
21:10 < sfix> can get a few more set of eyes on it
```

**Comment by garyttierney** [comment]
Looks good so far. Only real missing piece of the puzzle is finding out what direction a Mob is currently facing.

**Comment by stevesoltys** [comment]
@WizardJesse1 I think that's an issue with the A* pathfinding implementation, rather than this plugin.

**Comment by WizardJesse1** [comment]
@stevesoltys alright then, other then that it seems pretty good

File: data/plugins/follow/util.rb [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+  # An array containing the north, east, south, and west directions.
+  HORIZONTAL_DIRECTIONS = [Direction::NORTH, Direction::EAST, Direction::SOUTH, Direction::WEST]
+
+  ##
+  # A pathfinder used for simple following (i.e.: An NPC attacking a player).
+  SIMPLE_PATH_FINDER = SimplePathfindingAlgorithm.new($world)
+
+  ##
+  # A pathfinder used for precise following (i.e.: A player attacking another player).
+  FOLLOW_PATH_FINDER = AStarPathfindingAlgorithm.new($world, ChebyshevHeuristic.new)
```
Comment (garyttierney) [inline data/plugins/follow/util.rb]: Use the `EuclideanHeuristic` implementation (or fix Chebyshev)

File: data/plugins/follow/util.rb (hunk 3 above)
Comment (stevesoltys) [inline data/plugins/follow/util.rb]: Currently, attributes are limited to certain data types. See `AttributeType`.

Any other suggestions? @garyttierney

**Comment by stevesoltys** [comment]
Made most changes @garyttierney, only the collision issue remains.

Following single tile mobs seems perfect, besides the inability to "dance". Please confirm that it works just as well on your end @WizardJesse1

**Comment by WizardJesse1** [comment]
@stevesoltys it still seems to be holding diagonal http://recordit.co/TqilAsR40X it doesn't happen all the time but it does happen

File: data/plugins/follow/util.rb (hunk 3 above)
Comment (garyttierney) [inline data/plugins/follow/util.rb]: @stevesoltys something similar to combat maybe https://github.com/apollo-rsps/apollo/pull/260/files#diff-70a326256cf9b797a8a087b495c01987R3

**Comment by garyttierney** [comment]
@WizardJesse1 this might be down to not trying to get behind the player. That's still one thing to do IIRC.

**Comment by WizardJesse1** [comment]
o kk then @garyttierney

**Comment by garyttierney** [comment]
Latest changes look good. However, would still like the 'get behind' / dancing behavior in. Is this something you can handle @stevesoltys ?

**Comment by stevesoltys** [comment]
@garyttierney, going to work on getting that behavior in right now.  Also, mobs that are larger than one tile still need to be tested (and looking back at it now, they probably do not currently work).

**Comment by stevesoltys** [comment]
[https://cdn.streamable.com/video/mp4/0npo.mp4](url)

Will be committing these changes soon.

**Comment by Lmctruck30** [comment]
Issue with this is doors now. Doors are now blocking the walking path when opened. Also in your Ruby you put to listen for a player action. As it stands it is listening for all player actions not just the follow. So if you trade someone you will follow.

**Comment by garyttierney** [comment]
World#intersects() is a hack and incorrect, it doesn't correctly account for
paths that don't have equal X and Y deltas.

The best thing to do here is register listeners with a Region which can update
the CollisionMatrix when a door is opened / closed (though, one listener
should be able to do this for any type of dynamic object).

I'm going to look at some of this over the festive period.

**Comment by Lmctruck30** [comment]
@garyttierney Thank you. Be awesome to see this system finished =D

**Comment by Lmctruck30** [comment]
@garyttierney Is there an update on this PR?

**Comment by garyttierney** [comment]
@Lmctruck30 I've done some more work on this to accommodate for dynamic objects.  I need to wire up the new collision matrix code to some region listeners for adding / removing objects and it should be all set.

**Comment by Lmctruck30** [comment]
@garyttierney let me know if you want me to test.

**Comment by garyttierney** [comment]
Closed by https://github.com/apollo-rsps/apollo/commit/e5a6638e2fc0bc4f524906639c42893998c4e594.

**Pull request closed without merging**